kafka/client_async.py

```diff
@@ -512,46 +527,40 @@ def poll(self, timeout_ms=None, future=None, delayed_tasks=True):
         Returns:
             list: responses received (can be empty)
         """
-        if timeout_ms is None:
+        if future is not None:
+            timeout_ms = 100
```

Comment:
Yes, deserves a comment. The issue is that we now have to account for the future being resolved in a different thread. Prior to this change we would block on network IO and then check the future's completion after processing the network IO and any corresponding futures. But now that we have separated future handling from network IO, and we have (intentionally) moved the future handling into a separate section that is not locked, we can have a situation where one thread is called with a simple timeout and another is called with a future (block until resolved). The first thread acquires the lock and receives the response that will resolve the future in its _poll() call. The response is put on the internal pending_completions queue and the lock is released. Now the second thread takes the lock and checks whether the future is resolved. It is not, so it calls _poll with a full request timeout, which is now 5 minutes. The first thread continues, without the lock, and begins processing pending_completions. The future is resolved with the queued response, and the first thread finishes. But the second thread is now waiting for network IO that it thinks is necessary to complete its future. Except that the future is already done and so there is no network IO coming. It will simply wait for 5 minutes and then timeout before rechecking the future and then finally returning.

My solution here is to simply reduce the network IO timeout when called with a future to reduce the unneeded block time when this occurs. Another solution might be to try to register any future that is "blocking" and call _wake_up() on when such a future is resolved. I decided against that approach because I think it is too complex and it is likely to add overhead to every response, not just ones for which some thread is blocking on a future.

Reply:
I am glad I asked. I never would have realized that.

Reply:
Same =)